lib/Dialect/AIG/AIGOps.cpp

```diff
@@ -0,0 +1,141 @@
+//===- LoopScheduleOps.cpp - LoopSchedule CIRCT Operations ------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implement the AIG ops.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/AIG/AIGOps.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "mlir/IR/PatternMatch.h"
+
+using namespace mlir;
+using namespace circt;
+using namespace circt::aig;
+
+#define GET_OP_CLASSES
+#include "circt/Dialect/AIG/AIG.cpp.inc"
+
+OpFoldResult AndInverterOp::fold(FoldAdaptor adaptor) {
+  // TODO: Implement this.
+  return {};
+}
+
+LogicalResult AndInverterOp::canonicalize(AndInverterOp op,
+                                          PatternRewriter &rewriter) {
+  // TODO: Extend to more than 2 inputs.
+  if (op.getInputs().size() == 2) {
+    if ((op.getInputs()[0] == op.getInputs()[1]) &&
+        (op.isInverted(0) == op.isInverted(1))) {
```

Comment:
Flag check is missing.

```suggestion
        (op.isInverted(0) == op.isInverted(1)) && op.isInverted(0)) {
```
Also we need to implement three cases:
```
a & a -> a
not a & a -> 0
not a & not a -> not a
```

Reply:
Maybe this could iterate over the operands in a loop and keep a small map on the side with operands previously seen. And then when you see a duplicate you either discard id (if it's the same inversion), or you replace the gate with cosntant zero.

```
SmallDenseMap<Value, bool> seen;
SmallVector<Value> uniqueValues;
SmallVector<bool> uniqueInverts;
for (auto [value, inverted] : llvm::zip(op.getInputs(), op.getInverted())) {
  auto it = seen.find(value);
  if (it == seen.end()) {
    seen.insert(it, {value, inverted});
    uniqueValues.push_back(value);
    uniqueInverts.push_back(inverted);
  } else if (it->second != inverted) {
    // replace with const 0
    return
  }
}
// build new op with reduced input values
```

Reply:
Great point, thanks!